internal/app/machined/pkg/system/services/registry/registry.go

```diff
@@ -0,0 +1,196 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package registry provides a simple container registry service.
+package registry
+
+import (
+	"bytes"
+	"cmp"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/distribution/reference"
+	"github.com/opencontainers/go-digest"
+	"github.com/siderolabs/gen/xerrors"
+	"go.uber.org/zap"
+)
+
+// NewService creates a new instance of the registry service.
+func NewService(root string, logger *zap.Logger) *Service {
+	return &Service{root: root, logger: logger}
+}
+
+// Service is a container registry service.
+type Service struct {
+	logger *zap.Logger
+	root   string
+}
+
+// Run is an entrypoint to the API service.
+func (s *Service) Run(ctx context.Context) error {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc("GET /v2/{args...}", s.serveHTTP)
+
+	giveOk := func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(http.StatusOK) }
+	for _, p := range []string{"v2", "healthz"} {
+		mux.HandleFunc("GET /"+p, giveOk)
```

Comment:
I'm not sure if `mux` handles `HEAD` transparently, but I believe they do `HEAD /v2` requests as well